File: python/questions/longestBalanced.py
class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        ans = 0
        n = len(nums)
        for s in range(n):
            nst = set([nums[s]])
            d = 1 if nums[s] & 1 else -1
            for e in range(s+1, n):
                cn = nums[e]
                if cn not in nst:
                    nst.add(cn)
                    d += 1 if cn & 1 else -1
                if d == 0:
                    ans = max(ans, e-s+1)
        return ans

File: cpp/questions/longestBalanced.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int longestBalanced(vector<int>& nums) {
        int ans = 0;
        for (int s=0; s<nums.size(); s++) {
            std::set<int> nst = {nums[s]};
            int d = nums[s] & 1 ? 1 : -1;
            for (int e=s+1; e<nums.size(); e++) {
                auto it = nst.find(nums[e]);
                if (it == nst.end()) {
                    d += nums[e] & 1 ? 1 : -1;
                    nst.insert(nums[e]);
                }
                if (d == 0) {
                    ans = std::max(ans, e-s+1);
                }
            }
        }
        return ans;
    }
};
